nixos/modules/tasks/filesystems/zfs.nix

```diff
@@ -517,9 +522,18 @@ in
               done
               poolImported "${pool}" || poolImport "${pool}"  # Try one last time, e.g. to import a degraded pool.
               if poolImported "${pool}"; then
-                ${optionalString cfgZfs.requestEncryptionCredentials ''
+                ${optionalString (if isBool cfgZfs.requestEncryptionCredentials
+                                  then cfgZfs.requestEncryptionCredentials
+                                  else cfgZfs.requestEncryptionCredentials != []) ''
                   ${packages.zfsUser}/sbin/zfs list -rHo name,keylocation ${pool} | while IFS=$'\t' read ds kl; do
-                    (case "$kl" in
+                    (${if isBool cfgZfs.requestEncryptionCredentials
+                       then "" # requestEncryptionCredentials was true, so we loop over all (we insert no check)
+                       else ''
+                         if ! echo '${concatStringsSep cfgZfs.requestEncryptionCredentials "\n"}' | grep -qx "$ds"; then
+                           continue
+                         fi
+                       ''}
+                    case "$kl" in
```

Comment:
I haven't tested this bit yet, because I only have a single pool.  I plan to do this in the next few days in a VM.

Reply:
I have tested this in a VM with one extra pool and it works well.  I've also had a look at `nixos/tests/zfs.nix`, whether there is an easy way to test this functionality but it doesn't look like it is possible to perform reboots of the test VMs.

Reply:
It is actually possible https://github.com/NixOS/nixpkgs/blob/55079328c219fb53f3120a34fdba3a80d3be601b/nixos/tests/consul.nix#L161 

Do you want to implement that?

Reply:
There is `shutdown()` as well as `start()`.